x-pack/metricbeat/module/mssql/log/log.go

```diff
@@ -0,0 +1,223 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package db
+
+import (
+	"database/sql"
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/x-pack/metricbeat/module/mssql"
+)
+
+type dbInfo struct {
+	id   int
+	name string
+}
+
+type logSpace struct {
+	id                             int
+	totalLogSizeInBytes            int
+	usedLogSpaceInBytes            int
+	usedLogSpaceInPercent          float64
+	logSpaceInBytesSinceLastBackup int
+}
+
+type logStats struct {
+	databaseID            int
+	sizeMB                float64
+	activeSizeMB          float64
+	backupTime            string
+	sinceLastBackupMB     *float64
+	sinceLastCheckpointMB float64
+	recoverySizeMB        float64
+}
+
+func init() {
+	mb.Registry.MustAddMetricSet("mssql", "log", New,
+		mb.DefaultMetricSet(),
+		mb.WithHostParser(mssql.HostParser))
+}
+
+// MetricSet type defines all fields of the MetricSet
+type MetricSet struct {
+	mb.BaseMetricSet
+	log *logp.Logger
+	db  *sql.DB
+}
+
+// New create a new instance of the MetricSet
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The mssql log metricset is beta.")
+
+	logger := logp.NewLogger("mssql.log").With("host", base.HostData().SanitizedURI)
+
+	db, err := mssql.NewConnection(base.HostData().URI)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create connection to db")
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		log:           logger,
+		db:            db,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right format
+// It returns the event which is then forward to the output. In case of an error, a
+// descriptive error must be returned.
+func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
+	dbs, err := m.getDbsNames()
+	if err != nil {
+		reporter.Error(err)
+		return
+	}
+
+	for _, db := range dbs {
+		moduleFields := common.MapStr{
+			"database": common.MapStr{
+				"id":   db.id,
+				"name": db.name,
+			},
+		}
+		metricsetFields := common.MapStr{}
+
+		spaceUsage, err := m.getLogSpaceUsageForDb(db.name)
+		if err != nil {
+			reporter.Error(err)
+		} else {
+			metricsetFields["space_usage"] = spaceUsage
+		}
+
+		stats, err := m.getLogStats(db)
+		if err != nil {
+			reporter.Error(err)
```

Comment:
Also, if some of these errors happen, do we want to continue?

Reply:
Ok, you are right, I'll add some context. About the second question: Yes, I'd say that we want to continue. It's very unlikely that one of the objects is going to fail. Almost impossible that two of them will fail and it's not because of a network error (that should _explode_ somewhere else).

What I can do here is to handle the situation where `space_usage` and `stats` aren't filled to not try to report an empty event.

Reply:
Ok, sounds good.

Reply:
Checked the content of the event to not send it empty.